Kernel/Net/NE2000NetworkAdapter.cpp

```diff
@@ -0,0 +1,487 @@
+ * the PCI bus, including:
+ * - Realtek's RTL8029
+ * - VIA Technologies, Inc.'s VT86C926
+ *
+ * Official documentation from National Semiconductor includes:
+ *  - Datasheet "DP8390D/NS32490D NIC Network Interface Controller"
+ *  - Application Note 874 "Writing Drivers for the DP8390 NIC Family of Ethernet Controllers"
+ *
+ * This driver supports only the PCI variant.
+ *
+ * Remember, friends don't let friends use NE2000 network cards :^)
+ */
+
+// Page 0 registers
+#define REG_RW_COMMAND 0x00
+#define BIT_COMMAND_STOP (0b1 << 0)
+#define BIT_COMMAND_START (0b1 << 1)
+#define BIT_COMMAND_TXP (0b1 << 2)
+#define BIT_COMMAND_DMA_READ (0b001 << 3)
+#define BIT_COMMAND_DMA_WRITE (0b010 << 3)
+#define BIT_COMMAND_DMA_SEND (0b011 << 3)
+#define BIT_COMMAND_DMA_ABORT (0b100 << 3)
+#define BIT_COMMAND_DMA_FIELD (0b111 << 3)
+#define BIT_COMMAND_PAGE1 (0b01 << 6)
+#define BIT_COMMAND_PAGE2 (0b10 << 6)
+#define BIT_COMMAND_PAGE_FIELD (0b11 << 6)
+
+#define REG_WR_PAGESTART 0x01
+#define REG_WR_PAGESTOP 0x02
+#define REG_RW_BOUNDARY 0x03
+#define REG_RD_TRANSMITSTATUS 0x04
+#define REG_WR_TRANSMITPAGE 0x04
+#define REG_RD_NCR 0x05
+#define REG_WR_TRANSMITBYTECOUNT0 0x05
+#define REG_WR_TRANSMITBYTECOUNT1 0x06
+#define REG_RW_INTERRUPTSTATUS 0x07
+#define REG_RD_CRDMA0 0x08
+#define REG_WR_REMOTESTARTADDRESS0 0x08
+#define REG_RD_CRDMA1 0x09
+#define REG_WR_REMOTESTARTADDRESS1 0x09
+#define REG_WR_REMOTEBYTECOUNT0 0x0a
+#define REG_WR_REMOTEBYTECOUNT1 0x0b
+
+#define REG_RD_RECEIVESTATUS 0x0c
+#define BIT_RECEIVESTATUS_PRX (0b1 << 0)
+#define BIT_RECEIVESTATUS_CRC (0b1 << 1)
+#define BIT_RECEIVESTATUS_FAE (0b1 << 2)
+#define BIT_RECEIVESTATUS_FO (0b1 << 3)
+#define BIT_RECEIVESTATUS_MPA (0b1 << 4)
+
+#define REG_WR_RECEIVECONFIGURATION 0x0c
+#define BIT_RECEIVECONFIGURATION_SEP (0b1 << 0)
+#define BIT_RECEIVECONFIGURATION_AR (0b1 << 1)
+#define BIT_RECEIVECONFIGURATION_AB (0b1 << 2)
+#define BIT_RECEIVECONFIGURATION_AM (0b1 << 3)
+#define BIT_RECEIVECONFIGURATION_PRO (0b1 << 4)
+#define BIT_RECEIVECONFIGURATION_MON (0b1 << 5)
+
+#define REG_RD_FAE_TALLY 0x0d
+
+#define REG_WR_TRANSMITCONFIGURATION 0x0d
+#define BIT_WR_TRANSMITCONFIGURATION_LOOPBACK (0b10 << 0)
+
+#define REG_RD_CRC_TALLY 0x0e
+
+#define REG_WR_DATACONFIGURATION 0x0e
+#define BIT_DATACONFIGURATION_WTS (0b1 << 0)
+#define BIT_DATACONFIGURATION_BOS (0b1 << 1)
+#define BIT_DATACONFIGURATION_LS (0b1 << 2)
+#define BIT_DATACONFIGURATION_FIFO_8B (0b10 << 5)
+
+#define REG_RD_MISS_PKT_TALLY 0x0f
+
+#define REG_WR_INTERRUPTMASK 0x0f
+#define BIT_INTERRUPTMASK_PRX (0b1 << 0)
+#define BIT_INTERRUPTMASK_PTX (0b1 << 1)
+#define BIT_INTERRUPTMASK_RXE (0b1 << 2)
+#define BIT_INTERRUPTMASK_TXE (0b1 << 3)
+#define BIT_INTERRUPTMASK_OVW (0b1 << 4)
+#define BIT_INTERRUPTMASK_CNT (0b1 << 5)
+#define BIT_INTERRUPTMASK_RDC (0b1 << 6)
+#define BIT_INTERRUPTMASK_RST (0b1 << 7)
+
+#define REG_RW_IOPORT 0x10
+
+// Page 1 registers
+#define REG_RW_PHYSICALADDRESS0 0x01
+#define REG_RW_CURRENT 0x07
```

Comment:
How about we make all these things `static constexpr` instead of macros?
It seems nice to be explicit about the types used. :)